src/ngx_http_lua_semaphore.c

```diff
@@ -0,0 +1,523 @@
+
+    block = ngx_alloc(n, ngx_cycle->log);
+
+    if (block == NULL) {
+        return NULL;
+    }
+
+    dd("alloc semaphore block:%p", block);
+
+    mm->cur_epoch++;
+    mm->total += mm->num_per_block;
+    mm->used++;
+
+    block->mm = mm;
+    block->epoch = mm->cur_epoch;
+
+    sem = (ngx_http_lua_semaphore_t *) (block + 1);
+    sem->block = block;
+    sem->block->used = 1;
+
+    for (iter = sem + 1, i = 1; i < mm->num_per_block; i++, iter++) {
+        iter->block = block;
+        ngx_queue_insert_tail(&mm->free_queue, &iter->chain);
+    }
+
+    dd("ngx_http_lua_alloc_semaphore sem: %p", sem);
+
+    return sem;
+}
+
+
+void
+ngx_http_lua_cleanup_semaphore_mm(void *data)
+{
+    ngx_http_lua_semaphore_t            *sem;
+    ngx_http_lua_semaphore_t            *iter;
+    ngx_uint_t                           i;
+    ngx_http_lua_main_conf_t            *lmcf;
+    ngx_queue_t                         *q;
+    ngx_http_lua_semaphore_mm_block_t   *block;
+    ngx_http_lua_semaphore_mm_t         *mm;
+
+    lmcf = (ngx_http_lua_main_conf_t*) data;
+    mm = lmcf->semaphore_mm;
+
+    while (!ngx_queue_empty(&mm->free_queue)) {
+
+        q = ngx_queue_head(&mm->free_queue);
+        sem = ngx_queue_data(q, ngx_http_lua_semaphore_t, chain);
+        block = sem->block;
+
+        if (block->used == 0) {
+            iter = (ngx_http_lua_semaphore_t *) (block + 1);
+
+            for (i = 0; i < block->mm->num_per_block; i++, iter++) {
+                ngx_queue_remove(&iter->chain);
+            }
+
+            dd("free semaphore block: %p at final", block);
+
+            ngx_free(block);
+
+        } else {
+            ngx_log_error(NGX_LOG_ALERT, ngx_cycle->log, 0,
+               "ngx_http_lua_cleanup_semaphore_mm when cleanup"
+               " block: %p is still using by someone", block);
+            /* some thing go wrong just return directly */
+            return;
+        }
+    }
+
+    dd("ngx_http_lua_cleanup_semaphore_mm");
+}
+
+
+static void
+ngx_http_lua_free_semaphore(ngx_http_lua_semaphore_t *sem)
+{
+    ngx_http_lua_semaphore_t          *iter;
+    ngx_int_t                          start_epoch;
+    ngx_uint_t                         i;
+    ngx_http_lua_semaphore_mm_block_t *block;
+    ngx_http_lua_semaphore_mm_t       *mm;
+
+    block = sem->block;
+    block->used--;
+
+    mm = block->mm;
+    mm->used--;
+
+    start_epoch = mm->cur_epoch - mm->total/mm->num_per_block;
+
+    if (sem->block->used < (mm->num_per_block >> 1)
+        && block->epoch <= start_epoch + ((mm->cur_epoch - start_epoch) >> 1))
+    {
+        ngx_queue_insert_tail(&mm->free_queue, &sem->chain);
+
+    } else {
+        ngx_queue_insert_head(&mm->free_queue, &sem->chain);
+    }
+
+    if (block->used == 0 && mm->used < (mm->total >> 1)
+        && block->epoch <= (start_epoch + ((mm->cur_epoch-start_epoch) >> 1)))
+    {
+        /* load <= 50% and it's the older */
+        iter = (ngx_http_lua_semaphore_t*) (block+1);
```

Comment:
Style: spaces needed around the binary operator `+`. Also need a space before the `*` char.